pgxn/neon/file_cache.c

```diff
@@ -449,6 +487,252 @@ lfc_init(void)
 #endif
 }
 
+static FileCacheStateEntry*
+lfc_get_state(size_t* n_entries)
+{
+	size_t max_entries = *n_entries;
+	size_t i = 0;
+	FileCacheStateEntry* fs;
+
+	if (lfc_maybe_disabled() || max_entries == 0)	/* fast exit if file cache is disabled */
+		return NULL;
+
+	fs = (FileCacheStateEntry*)palloc(sizeof(FileCacheStateEntry) * max_entries);
+
+	LWLockAcquire(lfc_lock, LW_SHARED);
+
+	if (LFC_ENABLED())
+	{
+		dlist_iter	iter;
+		dlist_reverse_foreach(iter, &lfc_ctl->lru)
+		{
+			FileCacheEntry *entry = dlist_container(FileCacheEntry, list_node, iter.cur);
+			memcpy(&fs[i].key, &entry->key, sizeof entry->key);
+			memcpy(fs[i].bitmap, entry->bitmap, sizeof entry->bitmap);
+			if (++i == max_entries)
+				break;
+		}
+		elog(LOG, "LFC: save state of %ld chunks", (long)i);
+	}
+
+	LWLockRelease(lfc_lock);
+
+	*n_entries = i;
+	return fs;
+}
+
+/*
+ * Prewarm LFC cache to the specified state.
+ *
+ * Prewarming can interfere with accesses to the pages by other backends. Usually access to LFC is protected by shared buffers: when Postgres
+ * is reading page, it pins shared buffer and enforces that only one backend is reading it, while other are waiting for read completion.
+ *
+ * But it is not true for prewarming: backend can fetch page itself, modify and then write it to LFC. At the
+ * same time `lfc_prewarm` tries to write deteriorated image of this page in LFC. To increase concurrency, access to LFC files (both read and write)
+ * is performed without holding locks. So it can happen that two or more processes write different content to the same location in the LFC file.
+ * Certainly we can not rely on disk content in this case.
+ *
+ * To solve this problem we use two flags in LFC entry: `prewarm_requested` and `prewarm_started`. First is set before prewarm is actually started.
+ * `lfc_prewarm` writes to LFC file only if this flag is set. This flag is cleared if any other backend performs write to this LFC chunk.
+ * In this case data loaded by `lfc_prewarm` is considered to be deteriorated and should be just ignored.
+ *
+ * But as far as write to LFC is performed without holding lock, there is no guarantee that no such write is in progress.
+ * This is why second flag is used: `prewarm_started`. It is set by `lfc_prewarm` when is starts writing page and cleared when write is completed.
+ * Any other backend writing to LFC should abandon it's write to LFC file (just not mark page as loaded in bitmap) if this flag is set.
+ * So neither `lfc_prewarm`, neither backend are saving page in LFC in this case - it is just skipped.
```

Comment:
I thought about something like this (make backend wait completion of block load by prewarmer.
But I consider such probability as relatively small (otherwise, if most of the bocks which are loaded by prewarm are in any case accessed by backends, then we do not need prewarm at all).

Also prewarmer and backend can use different LSNs. So there may be no sense to wait until prewarm load block completion - if we can not use it in any case.

You're set iii. Check LSNs data assumes that we need to store LSN for each block. It will add extra 8 bytes to each entry. I already wrote that size of LFC hash entry seems to be quite critical, so I prefer no to extend it's size. 

Your `LFC_PREWARM_STARTED` flag servers the same goal as my `prewarm_started` flag. The main difference is that in your approachm backend will wait prwarm completion while in my case - just skip this entry. Certainly second approach is much more simple while your approach can in some cases provide better performance. But I do not think that this difference will be noticeable in the assumption that probability of such collision is  small enough.

Reply:
> 
> I'm not sure I follow. We prewarm not just to get the data into LFC, but to have backends hit the LFC rather than having to access the pageserver.

Sorry, that I was unclear.
So, there can be three difference situations:
1. OLTP workload with small number of clients
2. OLTP workload with large number of clients
3. OLAP workload

In case 3) prewarm is not needed - OLAP queries do it much more efficient themselves.
My experiments shows that in case 2)  prewarm is also not needed: larger number of parallel backends can do prewarm much faster than separate prewarm worker.

So actually we need prewarm only in case 3 - where there are just few backends which perform OLTP queries (index scan). If database is larger,  it may take quite long time for this few backends to access the whole working set - fill LFC with data. All this time this backends and so client will suffer from large delays (because of `get page` latency). Only in this case expect prewarm can be useful. But as far as there is small number of backends and larger database, chances that this backends will conflict with prewarm are very small. 


This why spending much efforts for more efficient handling of this conflicts seems to be waste of time.


> 
> > Also prewarmer and backend can use different LSNs. So there may be no sense to wait until prewarm load block completion - if we can not use it in any case.
> 
> That's what the LSN check is for.

Storing 8 bytes per page (not per 1Mb chunk!) seems to be too expensive: with 1TB LFC we will have 128Mb just for storing LSNs. It is the same as we have no for shared buffers. 

> 
> > You're set iii. Check LSNs data assumes that we need to store LSN for each block. It will add extra 8 bytes to each entry. I already wrote that size of LFC hash entry seems to be quite critical, so I prefer no to extend it's size.
> 
> Apologies that I didn't make this clearer, but what I meant there is that the prewarmer should check there that the LwLsn for each prefetched page hasn't been invalidated yet. No need to store an LSN on the chunk itself, as we can pull that from the LwLsn cache.

Not quite clear to me how it can work. So prewarmer get page tp be prewarmed, get its lwlsn mark it as "prewarm started", release lock and start waiting for response from PS. At the same moment some backend can write this page, this pages is thrown aways and then once again retrieved by some backend. This backend will see "prewarming" flag and waits until prewarm is completed. Ok, prewar is completed. How to check now now that load version is what we need? We should remember request LSN somewhere ... but where? Or prewarmer can try check it itself by once again checking lwlsn. But it is once again racy because there is actually no sync between backends and prewarmer.

I have discussed this problem with Heikki. Hw first suggest to perform prewarming through shared buffers - in this case we have proper locking. But I do not want for prewarm to affect shared buffer, especially taken in account that they are small. And once again -  if probability of conflict is very small, it is enough to have quite conservative check here and no need to involve conditional vars, lwlsn, ... 

LwLSN checks are also not so trivial, taken in account that it is limited in size (entries can be evicted).